### PR TITLE
FPN - Add min/max size params for pre-processing.

### PR DESCRIPTION
**Pull request by 23pointsNorth**
Allow for the `min_size`, `max_size` parameter to be set when creating an FPN network. 
It is following the Faster-RCNN template.

### CODE REVIEW

**Comment by yuyu2172** [comment]
Thanks. Can you set the default values for `FasterRCNN` so that the tests won't fail?

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 1]
```diff
@@ -26,6 +26,9 @@ class FasterRCNN(chainer.Chain):
         head (Link): A link that has the same interface as
             :class:`~chainercv.links.model.fpn.Head`.
             Please refer to the documentation found there.
+        min_size (int): A preprocessing paramter for :meth:`prepare`. Please
+            refer to a docstring found for :meth:`prepare`.
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: How about this?
```
Please refer to :meth:`prepare`.
```
`docstring` will be confusing when it rendered by sphinx.

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py [hunk 2]
```diff
@@ -56,8 +57,8 @@ class FasterRCNNFPNResNet50(FasterRCNNFPNResNet):
        Feature Pyramid Networks for Object Detection. CVPR 2017
 
     Args:
-       n_fg_class (int): The number of classes excluding the background.
-       pretrained_model (string): The weight file to be loaded.
+        n_fg_class (int): The number of classes excluding the background.
+        pretrained_model (string): The weight file to be loaded.
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: Why did you change the indentation?

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py [hunk 3]
```diff
@@ -99,10 +103,10 @@ class FasterRCNNFPNResNet101(FasterRCNNFPNResNet):
...
 
     Args:
-       n_fg_class (int): The number of classes excluding the background.
-       pretrained_model (string): The weight file to be loaded.
-           This can take :obj:`'coco'`, `filepath` or :obj:`None`.
-           The default value is :obj:`None`.
+        n_fg_class (int): The number of classes excluding the background.
+        pretrained_model (string): The weight file to be loaded.
+            This can take :obj:`'coco'`, `filepath` or :obj:`None`.
+            The default value is :obj:`None`.
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: ditto.

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py [hunk 4]
```diff
@@ -19,7 +19,8 @@ class FasterRCNNFPNResNet(FasterRCNN):
     A subclass of this class should have :obj:`_base` and :obj:`_models`.
     """
 
-    def __init__(self, n_fg_class=None, pretrained_model=None):
+    def __init__(self, n_fg_class=None, pretrained_model=None,
+                 min_size=800, max_size=1333):
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: `min_size` and `max_size` are not used.

**Comment by Hakuyume** [comment]
> Thanks. Can you set the default values for FasterRCNN so that the tests won't fail?

Perhaps, we can modify the tests to pass those parameters explicitly.  In that case, we can remove the default value from `FasterRCNN`.

**Comment by Hakuyume** [comment]
> Thanks. Can you set the default values for FasterRCNN so that the tests won't fail?

Perhaps, we can modify the tests to pass those parameters explicitly.  In that case, we can remove the default value from `FasterRCNN`.

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 1 above)
Comment (23pointsNorth) [reply chainercv/links/model/fpn/faster_rcnn.py]: This was done in the same style as the faster-rcnn docs with regards to the param.

**Comment by Hakuyume** [comment]
@23pointsNorth Can you add tests that change `min_size` or `max_size`?
https://github.com/chainer/chainercv/blob/master/tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py#L120-L126

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py (hunk 2 above)
Comment (23pointsNorth) [reply chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: The rest of the code is indented with 4 space. 
These lines were with 3. I am happy to revert if comments don't follow this requirement.

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py (hunk 4 above)
Comment (23pointsNorth) [reply chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: That was a versioning mistake. It will be added on L39.

**Comment by 23pointsNorth** [comment]
I am not sure how you expect the testing to look like. If we inspect the other place `preprare` is used, i.e. https://github.com/chainer/chainercv/blob/master/tests/links_tests/model_tests/faster_rcnn_tests/test_faster_rcnn.py#L81 only the input param is checked against. 

I can alter the tests to include a value for min/max, such that we can remove the default in the class (but I think it may be unnecessary).

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py (hunk 4 above)
Comment (Hakuyume) [reply chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: That was my mistake. Thank you for fixing!

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py (hunk 2 above)
Comment (Hakuyume) [reply chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: This was my mistake. Thank you for fixing!

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 1 above)
Comment (Hakuyume) [reply chainercv/links/model/fpn/faster_rcnn.py]: I see.

**Comment by Hakuyume** [comment]
> I am not sure how you expect the testing to look like. If we inspect the other place preprare is used, i.e. https://github.com/chainer/chainercv/blob/master/tests/links_tests/model_tests/faster_rcnn_tests/test_faster_rcnn.py#L81 only the input param is checked against.

The test in `faster_rcnn_tests` checks the effect of `min_size/max_size` because it uses values that are different from the default values (`(200, 400)` vs `(600, 1000)`). On the other hand, the test in `fpn_tests` uses the default value. So the test cannot detect an error even if `FasterRCNN` ignores those parameters. At least we should use different values for the test. It would be better to check both default params and other params (with `testing.parameterize`).

> I can alter the tests to include a value for min/max, such that we can remove the default in the class (but I think it may be unnecessary, Edit based on:

I'm sorry for my confusing comment. It is OK to keep the default values. I just mentioned the side-effect.

**Comment by 23pointsNorth** [comment]
@Hakuyume I attempted to add extra tests. I changed them to follow the ones in [test_faster_rcnn](https://github.com/chainer/chainercv/blob/master/tests/links_tests/model_tests/faster_rcnn_tests/test_faster_rcnn.py#L74) in regards to `in_shape`, `expected_shape`. 

The `paramterize` tag becomes a bit clunky, because in our tests we want to make the permutations of 3*`n_fg_class`, 2*(`min_size`, `max_size`, `in_shape`, `expected shape`). I don't know if I can give 2 dict and the testing can do the permutations of them.

File: tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py [hunk 5]
```diff
@@ -31,24 +31,35 @@ def __call__(self, x):
...
         )
 
 
 @testing.parameterize(
-    {'n_fg_class': 1},
-    {'n_fg_class': 5},
-    {'n_fg_class': 20},
+    {'n_fg_class': 1, 'min_size': 200, 'max_size': 400, 
+     'in_shape': (3, 200, 50), 'expected_shape': (3, 400, 100)},
+    {'n_fg_class': 1, 'min_size': 800, 'max_size': 133,
```
Comment (Hakuyume) [inline tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py]: `133` is typo of `1333`?

File: tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py [hunk 6]
```diff
@@ -117,13 +128,19 @@ def test_predict_gpu(self):
...
-        x, scales = self.link.prepare(imgs)
-        self.assertEqual(x.shape, (2, 3, 800, 1088))
+    def check_prepare(self):
+        x = _random_array(np, self.in_shape)
+        out, scales = self.link.prepare(x)
+        self.assertIsInstance(out, np.ndarray)
+        self.assertEqual(out.shape, self.expected_shape)
+
+    def test_prepare_cpu(self):
+        self.check_prepare()
```
Comment (Hakuyume) [inline tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py]: We do not need to make `_cpu/_gpu` because `prepare` is not related `cupy`.

**Comment by Hakuyume** [comment]
You can use `testing.product_dict`.

```python
@testing.parameterize(*testing.product_dict(
    [
        {'n_fg_class': 1},
        {'n_fg_class': 5},
        {'n_fg_class': 20},
    ],
    [
        {
            'in_shape': (3, 480, 640),
            'min_size': 800, 'max_size': 1333,
            'expected_shape': (3, 800, 1088),
        },
        {
            'in_shape': (3, 200, 50),
            'min_size': 200, 'max_size': 400,
            'expected_shape': (3, 400, 100),
        },
    ],
))
class TestFasterRCNN(unittest.TestCase):
```

File: tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py [hunk 7]
```diff
@@ -31,24 +31,35 @@ def __call__(self, x):
...
 
 
 @testing.parameterize(
-    {'n_fg_class': 1},
-    {'n_fg_class': 5},
-    {'n_fg_class': 20},
+    {'n_fg_class': 1, 'min_size': 200, 'max_size': 400, 
+     'in_shape': (3, 200, 50), 'expected_shape': (3, 400, 100)},
+    {'n_fg_class': 1, 'min_size': 800, 'max_size': 133, 
+     'in_shape': (3, 480, 640), 'expected_shape': (3, 800, 1088)},
```
Comment (Hakuyume) [inline tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py]: `prepare` can accept a batch of images of different shapes. 
How about `'in_sizes': [(480, 640), (320, 320)], 'expected_size': (800, 1088)` ? (I omit `3` because it is not parameterized)

File: tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py (hunk 5 above)
Comment (23pointsNorth) [reply tests/links_tests/model_tests/fpn_tests/test_faster_rcnn.py]: Yes, a typo.

**Comment by 23pointsNorth** [comment]
I think I've incorporated all of the requests. A general **unexpected behaviour** I found based on the tests i.e. 
https://travis-ci.org/chainer/chainercv/jobs/505234966#L847
If the `max_size` is the parameter based on which the size is calculated, but it is not divisible by `stride`, based on line 
https://github.com/chainer/chainercv/blob/master/chainercv/links/model/fpn/faster_rcnn.py#L165
we'll get a resulting image size that is larger than `max_size` (i.e. max_size of 400 -> final image size 416).

**Comment by Hakuyume** [comment]
> If the max_size is the parameter based on which the size is calculated, and if it is not divisible by stride, based on line
https://github.com/chainer/chainercv/blob/master/chainercv/links/model/fpn/faster_rcnn.py#L165
we'll get a resulting image size that is larger than max_size (i.e. max_size of 400 -> final image size 416).

Actually, this is an expected behavior (same as the Detecton's implementation).
But I agree with you that it is confusing because the output can exceed `max_size`.

> This I think is ok in general, but should be documented somewhere.

How about adding a note here?
https://github.com/chainer/chainercv/blob/f1ad965c005954c9b19c49c5e5741712b070feb5/chainercv/links/model/fpn/faster_rcnn.py#L31
For example,
```
max_size (int): A preprocessing paramter for :meth:`prepare`. 
        Note that the result of :meth:`prepare` can exceed this size
        because of alignment.
```
